File: src/file_suffix_inspector.js
const core = require('@actions/core');
const github = require('@actions/github');
const fs = require('fs');
const path = require('path');

async function run() {
    try {
        const suffixesRaw = core.getInput('suffixes');
        const suffixes = suffixesRaw ? suffixesRaw.split(',').map(s => s.trim()) : [];
        const includeDirectoriesRaw = core.getInput('include_directories') || "src/test/";
        const includeDirectories = includeDirectoriesRaw ? includeDirectoriesRaw.split(',').map(s => s.trim()) : [];
        const excludeDirectoriesRaw = core.getInput('exclude_directories') || "dist,node_modules,coverage,target,.idea,.github";
        const excludeDirectories = excludeDirectoriesRaw ? excludeDirectoriesRaw.split(',').map(s => s.trim()) : [];
        const excludeFilesRaw = core.getInput('exclude_files');
        const excludeFiles = excludeFilesRaw ? excludeFilesRaw.split(',').map(s => s.trim()) : [];
        const caseSensitivityInput = core.getInput('case_sensitivity');
        const caseSensitivity = caseSensitivityInput ? caseSensitivityInput === 'true' : true;
        const logicInput = core.getInput('logic');
        const logic = logicInput ? logicInput === 'true' : true;
        const reportFormat = core.getInput('report_format') || "console";
        const verboseLogging = core.getInput('verbose_logging') === 'true';
        const failOnViolations = core.getInput('fail_on_violations') === 'true';

        if (verboseLogging) {
            core.info(`Suffixes: ${JSON.stringify(suffixes)}`);
            core.info(`Include directories: ${JSON.stringify(includeDirectories)}`);
            core.info(`Exclude directories: ${JSON.stringify(excludeDirectories)}`);
            core.info(`Exclude files: ${JSON.stringify(excludeFiles)}`);
            core.info(`Case sensitivity: ${caseSensitivity}`);
            core.info(`Logic: ${logic}`);
            core.info(`Report format: ${reportFormat}`);
            core.info(`Fail on violations: ${failOnViolations}`);
        }

        let violationsCount = 0;
        const violations = [];

        function scanDirectory(directory, level= 1) {
            const files = fs.readdirSync(directory, { withFileTypes: true });

            for (const file of files) {
                const fullPath = path.join(directory, file.name);
                const filename = file.name;

                if (file.isDirectory()) {
                    // check exclude root directories
                    if (level === 1 && excludeDirectories.includes(file.name)) {
                        if (verboseLogging) {
                            core.info(`Skipping excluded directory: ${file.name}`);
                        }
                    } else {
                        scanDirectory(fullPath, level + 1);
                    }
                } else {
                    for (const includeDir of includeDirectories) {
                        // check included directories
                        if (fullPath.includes(includeDir)) {
                            if (excludeFiles.includes(filename)) {
                                if (verboseLogging) {
                                    core.info(`Excluded file: ${fullPath}`);
                                }

                                continue;
                            }

                            const checkFilenameWithExtension = caseSensitivity ? filename : filename.toLowerCase();
                            const checkFilename = checkFilenameWithExtension.split(".")[0];
                            const matchesSuffix = suffixes.some(suffix => {
                                const checkSuffix = caseSensitivity ? suffix : suffix.toLowerCase();
                                return logic ? checkFilename.endsWith(checkSuffix) : checkFilename.includes(checkSuffix);
                            });

                            if (!matchesSuffix) {
                                if (verboseLogging) {
                                    core.info(`Violating file: ${fullPath}`);
                                }
                                violationsCount++;
                                violations.push(fullPath);
                            } else {
                                if (verboseLogging) {
                                    core.info(`Valid file: ${fullPath}`);
                                }
                            }
                        }
                    }
                }
            }
        }

        scanDirectory(path.join(process.cwd()));
        // TODO - test moznosti nastaveni root dir zde

        if (verboseLogging) {
            core.info(`Total violations: ${violationsCount}`);
            core.info(`Violating files: ${violations.join(', ')}`);
        }

        core.setOutput('conventions_violations', violationsCount);

        if (reportFormat === 'console' || verboseLogging) {
            console.log(`Total violations: ${violationsCount}`);
            console.log(`Violating files: ${violations.join(', ')}`);
        } else if (reportFormat === 'csv') {
            const csvContent = violations.join('\n');
            fs.writeFileSync('violations.csv', csvContent);
            core.setOutput('report_file', 'violations.csv');
        } else if (reportFormat === 'json') {
            const jsonContent = JSON.stringify({ violations });
            fs.writeFileSync('violations.json', jsonContent);
            core.setOutput('report_file', 'violations.json');
        }

        if (failOnViolations && violationsCount > 0) {
            core.setFailed(`There are ${violationsCount} test file naming convention violations.`);
        }
    } catch (error) {
        console.log(`Action failed with error: ${error}`);
        core.setFailed(`Action failed with error: ${error}`);
    }
}

module.exports.run = run;

if (require.main === module) {
    run();
}

File: src/file_suffix_inspector.py
import os
import json
import csv
from pathlib import Path


def get_input(name: str) -> str:
    value = os.getenv(f'INPUT_{name.upper()}')
    return value


def set_output(name: str, value: str):
    output_file = os.getenv('GITHUB_OUTPUT', None)
    if output_file:
        with open(output_file, 'a') as f:
            f.write(f'{name}={value}\n')
    else:
        raise EnvironmentError('GITHUB_OUTPUT environment variable is not set.')


def set_failed(message: str):
    print(f'::error::{message}')
    exit(1)


def info(message: str):
    print(message)


def run():
    try:
        suffixes_raw = get_input('suffixes')
        suffixes = suffixes_raw.split(',') if suffixes_raw else []
        include_directories_raw = get_input('include_directories')
        include_directories = include_directories_raw.split(',')
        excludes_raw = get_input('excludes')
        excludes = excludes_raw.split(',')
        case_sensitive = get_input('case_sensitive') == 'true'
        contains = get_input('contains') == 'true'
        report_format = get_input('report_format')
        verbose_logging = get_input('verbose_logging') == 'true'
        fail_on_violation = get_input('fail_on_violation') == 'true'

        if verbose_logging:
            info(f'Suffixes: {suffixes}')
            info(f'Include directories: {include_directories}')
            info(f'Excludes: {excludes}')
            info(f'Case sensitivity: {case_sensitive}')
            info(f'Contains: {contains}')
            info(f'Report format: {report_format}')
            info(f'Fail on violations: {fail_on_violation}')

        violation_count = 0
        violations = []

        def scan_directory(directory: Path, level: int = 1):
            if verbose_logging:
                info(f'Scanning directory: {directory}')

            nonlocal violation_count
            nonlocal violations
            for item in directory.iterdir():
                if item.is_dir():
                    scan_directory(item, level + 1)
                else:
                    for include_dir in include_directories:
                        # Check if item path contains the include directory
                        if str(item).__contains__(include_dir):
                            # Check if item is defined as an exception to exclude
                            if item.name in excludes:
                                if verbose_logging:
                                    info(f'Excluded file: {item}')
                                continue

                            check_filename_with_extension = item.name if case_sensitive else item.name.lower()
                            check_filename = check_filename_with_extension.split(".")[0]
                            matches_suffix = any(
                                (check_filename.endswith(suffix) if contains else suffix in check_filename)
                                for suffix in (suffixes if case_sensitive else map(str.lower, suffixes))
                            )

                            if not matches_suffix:
                                if verbose_logging:
                                    info(f'Violating file: {item}')
                                violation_count += 1
                                violations.append(str(item))
                            else:
                                if verbose_logging:
                                    info(f'Valid file: {item}')

        scan_directory(Path.cwd())

        set_output('violation_count', str(violation_count))

        if report_format == 'console' or verbose_logging:
            print(f'Total violations: {violation_count}')
            print(f'Violating files: {violations}')

        if report_format == 'csv':
            with open('violations.csv', mode='w', newline='') as file:
                writer = csv.writer(file)
                writer.writerows([[violation] for violation in violations])
            set_output('report_path', 'violations.csv')
        elif report_format == 'json':
            with open('violations.json', mode='w') as file:
                json.dump({'violations': violations}, file)
            set_output('report_path', 'violations.json')

        if fail_on_violation and violation_count > 0:
            set_failed(f'There are {violation_count} test file naming convention violations.')

    except Exception as error:
        print(f'Action failed with error: {error}')
        set_failed(f'Action failed with error: {error}')


if __name__ == '__main__':
    run()
